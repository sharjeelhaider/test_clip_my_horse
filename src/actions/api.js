import {asyncActionNames, buildAsyncActions} from './utils/asyncUtils.js';
import {encode} from 'base-64';
// Constants
export const LOAD_HORSES = asyncActionNames('LOAD_HORSES');

// Actions
export const receiveHorses = buildAsyncActions(LOAD_HORSES);

// Async actions
export const fetchData = (
  username,
  password,
  endpoint,
  fetchAction,
) => async dispatch => {
  let auth = {Authorization: 'Basic ' + encode(username + ':' + password)};
  const requestOptions = {
    method: 'GET',
    headers: auth,
  };
  try {
    dispatch(fetchAction.request());
    const response = await fetch(endpoint, requestOptions);
    if (handleErrors(response)) {
      const responseJson = await response.json();
      dispatch(fetchAction.success(responseJson));
    }
  } catch (error) {
    dispatch(fetchAction.failure(error));
  }
};

function handleErrors(response) {
  if (!response.ok) {
    throw Error('Authentication error');
    //Api always return null for statusText
    // throw Error(response.statusText);
  }
  return response;
}
