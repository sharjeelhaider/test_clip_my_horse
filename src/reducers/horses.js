import {LOAD_HORSES} from '../actions/api';

export const initialState = {
  // Exporting it for test purposes
  error: null,
  isLoading: false,
  horses: null,
  isLoggedIn: false,
};

export default function horses(state = initialState, action) {
  switch (action.type) {
    case LOAD_HORSES.failure:
      return {...state, error: action.error, isLoading: false};
    case LOAD_HORSES.request:
      return {...state, error: null, isLoading: true};
    case LOAD_HORSES.success:
      return {
        ...state,
        error: null,
        isLoading: false,
        isLoggedIn: true,
        horses: action.data,
      };
    default:
      return state;
  }
}
