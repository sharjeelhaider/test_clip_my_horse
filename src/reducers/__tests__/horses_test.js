import horses from '../horses';
import {initialState} from '../horses';
import {receiveHorses} from '../../actions/api';
import data from '../../../config/jest/dataMock';

it('returns the same state on an unhandled action', () => {
  expect(horses(initialState, {type: 'HELLO_WORLD'})).toMatchSnapshot();
});

it('handles LOAD_HORSES_REQUEST action', () => {
  expect(horses(initialState, receiveHorses.request())).toMatchSnapshot();
});

it('handles LOAD_HORSES_FAILURE action', () => {
  const error = new Error('Look ma! I am an error');
  expect(horses(initialState, receiveHorses.failure(error))).toMatchSnapshot();
});

it('handles LOAD_HORSES_SUCCESS action', () => {
  // BOOM. If we change that specific reducer case, we'll compare the Snapshot results
  // and update them if the change is expected.
  expect(horses(initialState, receiveHorses.success(data))).toMatchSnapshot();
});
