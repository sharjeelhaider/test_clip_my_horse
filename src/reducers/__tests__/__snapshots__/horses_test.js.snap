// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`handles LOAD_HORSES_FAILURE action 1`] = `
Object {
  "error": [Error: Look ma! I am an error],
  "horses": null,
  "isLoading": false,
  "isLoggedIn": false,
}
`;

exports[`handles LOAD_HORSES_REQUEST action 1`] = `
Object {
  "error": null,
  "horses": null,
  "isLoading": true,
  "isLoggedIn": false,
}
`;

exports[`handles LOAD_HORSES_SUCCESS action 1`] = `
Object {
  "error": null,
  "horses": Array [
    Object {
      "id": "df579a48-a7ca-4afc-89d0-caabcfac87d7",
      "name": "American Beauty",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/51e87826c943fd896385734287bbac27fd43c74a.jpeg",
    },
    Object {
      "id": "8b1b98fe-077f-4faf-bb16-f088f3b6dfdf",
      "name": "Arabian Night",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/dae8d44c97b286e627a62569c2ff46bbb4575270.jpeg",
    },
    Object {
      "id": "49092d28-2d7b-4774-8a40-7c1a9d3e7634",
      "name": "Autumn Gold",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/b05678d2169739f280993fb8636ba017ff9b6787.jpeg",
    },
    Object {
      "id": "be11640d-1fe6-4d73-a664-0fd84ea6338c",
      "name": "Ancient Orange",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/5ea73acc5924c18bec341e3b30eb6ba8ea463d28.jpeg",
    },
    Object {
      "id": "a04e73cf-6ceb-40e9-969e-e94b02f4941e",
      "name": "Amazing Grace",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/7d860756e5c7fc9df00b39b459b868a12edbac9a.jpeg",
    },
    Object {
      "id": "a3c4c44c-1975-4d6d-a870-af9cd5d23628",
      "name": "Apple Candy",
      "picture": "https://d2vtwgzbrv2xnd.cloudfront.net/a7acc104b765cf17c4adeabaa2ca401ace08a5df.jpeg",
    },
  ],
  "isLoading": false,
  "isLoggedIn": true,
}
`;

exports[`returns the same state on an unhandled action 1`] = `
Object {
  "error": null,
  "horses": null,
  "isLoading": false,
  "isLoggedIn": false,
}
`;
